### PR TITLE
feat(cache): improved customizability

### LINKED ISSUE #2650: [Feature Request]: `hono/cache` cacheName should also take function
### What is the feature you are proposing?

In the `hono/cache` middleware, how can we add query params or other variables from the context? Like caching the data based on different `limit` parameters or user profiles/roles.

I checked the code and it's a simple implementation, I can create a PR for this 😀

### PR DESCRIPTION
**Pull request by MathurAditya724**
Closes #2650 

- [x] `cacheName` can accept a function, which returns a string
- [x] Added `keyGenerator` function, to define custom cache keys

### CODE REVIEW

**Comment by yusukebe** [comment]
Hey @MathurAditya724 

Good! I like using `((c: Context) => Promise<string> | string)`. There are some things that you want to do:

* Writing proper tests.
* `bun run denoify` to generate files for Deno and push them.
* `bun run format:fix && bun run lint:fix` to format the code.

Thanks!

**Comment by MathurAditya724** [comment]
Hey @yusukebe, I have completed all the requested changes kindly have a look. Regarding the test, I have added tests that seem reasonable to me but do let me know if you have any queries

**Comment by MathurAditya724** [comment]
I have added this [PR](https://github.com/honojs/website/pull/350) for the website, reflecting the behavior of the new API

File: src/test-utils/setup-vitest.ts [hunk 1]
```diff
@@ -28,6 +28,10 @@ class MockCache {
     return this.store.get(key) || null
   }
 
+  async keys() {
```
Comment (yusukebe) [inline src/test-utils/setup-vitest.ts:31]: Mocking `keys()` is good!

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
Hi @MathurAditya724 

Great! I'm glad you also created the PR for the website. I'll merge this to the `main` now and ship it with the next **patch** release since it's not such a big feature. Thank you for your contribution!

**Comment by kosiumeigbo** [comment]
thank you @MathurAditya724 . I have a use case for this and this has come in handy!

**Pull request merged**